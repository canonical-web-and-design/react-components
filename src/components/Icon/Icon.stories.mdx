import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs/blocks";

import Icon, { ICONS } from "./Icon";

import "./Icon.stories.scss";

<Meta title="Icon" component={Icon} />

### Icon

This is a [React](https://reactjs.org/) component for the Vanilla [Icon](https://docs.vanillaframework.io/patterns/icons/).

Icons provide visual context and enhance usability.

### Props

<ArgsTable of={Icon} />

### Default

<Canvas>
  <Story name="Default">
    <Icon name={ICONS.plus} />
    <Icon name={ICONS.minus} />
    <Icon name={ICONS.expand} />
    <Icon name={ICONS.collapse} />
    <Icon name={ICONS.spinner} />
    <Icon name={ICONS.drag} />
    <Icon name={ICONS.close} />
    <Icon name={ICONS.help} />
    <Icon name={ICONS.information} />
    <Icon name={ICONS.delete} />
    <Icon name={ICONS.externalLink} />
    <Icon name={ICONS.chevronUp} />
    <Icon name={ICONS.chevronDown} />
    <Icon name={ICONS.menu} />
    <Icon name={ICONS.code} />
    <Icon name={ICONS.copy} />
    <Icon name={ICONS.search} />
    <Icon name={ICONS.share} />
    <Icon name={ICONS.user} />
    <Icon name={ICONS.anchor} />
    <Icon name={ICONS.success} />
    <Icon name={ICONS.warning} />
    <Icon name={ICONS.error} />
  </Story>
</Canvas>

### Custom

To use custom icons that are not included in Vanilla you need to provide your own icon styling following the `.p-icon--{name}` convention.

<Canvas>
  <Story name="Custom">
    <Icon name="custom" />
  </Story>
</Canvas>
