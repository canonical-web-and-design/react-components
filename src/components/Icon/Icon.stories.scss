@import "~vanilla-framework/scss/settings";
@import "~vanilla-framework/scss/base_placeholders";
@include vf-b-placeholders;

@import "~vanilla-framework/scss/base_icon-definitions";
@import "~vanilla-framework/scss/patterns_icons";
@include vf-p-icons-common;

.p-icon--custom {
  @extend %icon;

  background-image: url(#{$assets-path}a4d31b28-icon-quote.svg);
}
