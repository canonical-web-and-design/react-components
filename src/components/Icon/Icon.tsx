import classNames from "classnames";
import PropTypes from "prop-types";
import type { HTMLProps } from "react";
import React from "react";

export enum ICONS {
  plus = "plus",
  minus = "minus",
  expand = "expand",
  collapse = "collapse",
  spinner = "spinner",
  drag = "drag",
  close = "close",
  help = "help",
  information = "information",
  delete = "delete",
  externalLink = "external-link",
  chevronDown = "chevron-down",
  chevronUp = "chevron-up",
  menu = "menu",
  code = "code",
  copy = "copy",
  search = "search",
  share = "share",
  user = "user",
  anchor = "anchor",
  success = "success",
  warning = "warning",
  error = "error",
}

type Props = {
  className?: string;
  name: ICONS | string;
} & HTMLProps<HTMLElement>;

/**
 * Icon
 *
 * @param name One of built-in Vanilla icons or a name of a custom icon that follows `p-icon--{name}` convention.
 * @returns Icon
 */
const Icon = ({ className, name, ...props }: Props): JSX.Element => (
  <i className={classNames(className, `p-icon--${name}`)} {...props} />
);

Icon.propTypes = {
  className: PropTypes.string,
  name: PropTypes.string.isRequired,
};

export default Icon;
