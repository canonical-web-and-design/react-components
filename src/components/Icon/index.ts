export { default, ICONS } from "./Icon";
