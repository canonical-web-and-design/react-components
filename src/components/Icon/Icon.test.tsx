import { shallow } from "enzyme";
import React from "react";

import Icon, { ICONS } from "./Icon";

describe("Icon", () => {
  it("renders", () => {
    const wrapper = shallow(<Icon name={ICONS.success} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("sets correct class name based on given name", () => {
    const wrapper = shallow(<Icon name="test" />);
    expect(wrapper.prop("className").includes("p-icon--test")).toBe(true);
  });
});
