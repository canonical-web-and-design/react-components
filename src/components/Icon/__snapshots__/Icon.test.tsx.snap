// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon renders 1`] = `
<i
  className="p-icon--success"
/>
`;
